File: PI3C/ejercicios/ejercicio1.c
/*
 * ejercicio1.c
 *
 *  Created on: 30 nov. 2020
 *      Author: LENOVO
 */

#include "ejercicio1.h"

bool ej1Recursivo(int a, int matriz[a][a]){
	return ej1RecAux(a, matriz, 0, 0, a);
}

bool ej1RecAux(int a, int matriz[a][a], int i, int j, int tam) {
	int maxFila = i + (tam-1);
	int maxColumna = j + (tam-1);
	set vertices = set_empty(int_type);
	set_add(&vertices, &matriz[i][j]);
	set_add(&vertices, &matriz[maxFila][j]);
	set_add(&vertices, &matriz[i][maxColumna]);
	set_add(&vertices, &matriz[maxFila][maxColumna]);
	bool res = set_size(&vertices)==4 || tam==1;
	if(tam%2==0) {
		if(res==true) {
			return ej1RecAux(a, matriz, i, j, tam/2)
					&& ej1RecAux(a, matriz, i+tam/2, j, tam/2)
					&& ej1RecAux(a, matriz, i, j+tam/2, tam/2)
					&& ej1RecAux(a, matriz, i+tam/2, j+tam/2, tam/2);
		}else {
			return res;
		}
	}else{
		return res;
	}
}

File: PI3Java/PI3OrdóñezMárquezJuanEmilio/src/ejercicios/Ejercicio1.java
package ejercicios;

import java.util.HashSet;
import java.util.Set;

public class Ejercicio1 {
	
	public static Boolean ej1Recursivo(Integer[][] matriz) {
		 return ej1RecAux(matriz, 0, 0, matriz.length);
	}
	
	private static Boolean ej1RecAux(Integer[][] matriz, int i, int j, int tam) {
		int maxFila = i + (tam-1);
		int maxColumna = j + (tam-1);
		Set<Integer> vertices = new HashSet<>();
		vertices.add(matriz[i][j]);
		vertices.add(matriz[maxFila][j]);
		vertices.add(matriz[i][maxColumna]);
		vertices.add(matriz[maxFila][maxColumna]);
		Boolean res = vertices.size()==4 || tam==1;
		if(tam%2==0) {
			if(res==true) {
				return ej1RecAux(matriz, i, j, tam/2)
						&& ej1RecAux(matriz, i+tam/2, j, tam/2)
						&& ej1RecAux(matriz, i, j+tam/2, tam/2)
						&& ej1RecAux(matriz, i+tam/2, j+tam/2, tam/2);
			}else {
				return res;
			}
		}else{
			return res;
		}
	}
	
}
